Add new case for copy_on_read attach disk

dzhengfy opened the pull request:
This implements two scenarios:
1. Coldplug/cold-unplug a disk with copy_on_read=on
2. Hotplug/hotunplug a disk with copy_on_read=on

Signed-off-by: Dan Zheng <dzheng@redhat.com>

dzhengfy commented:
```
# avocado run --vt-type libvirt --vt-machine-type q35 virtual_disks.multidisks..single_disk_test.disk_driver_option.option_copy_on_read_on
 (1/2) type_specific.io-github-autotest-libvirt.virtual_disks.multidisks.hotplug.single_disk_test.disk_driver_option.option_copy_on_read_on: PASS (82.28 s)
 (2/2) type_specific.io-github-autotest-libvirt.virtual_disks.multidisks.coldplug.single_disk_test.disk_driver_option.option_copy_on_read_on: PASS (60.83 s)
```

chunfuwen commented on libvirt/tests/cfg/virtual_disks/virtual_disks_multidisks.cfg:
```diff
@@ -372,22 +372,30 @@
                         - option_copy_on_read_on:
+                            hotplug:
+                                test_disk_option_cmd = "no"
+                                virt_disk_at_dt_disk = "yes"
```
variants:
        - hotplug:
            virt_disk_device_hotplug = "yes"
        - coldplug:
            virt_disk_device_hotplug = "no"
At end of cfg file, we have virt_disk_device_hotplug to diff cold or hot.  Do you really need introduce new variable virt_disk_at_dt_disk and virt_disk_cold_dt ?

chunfuwen commented on libvirt/tests/src/virtual_disks/virtual_disks_multidisks.py:
```diff
@@ -1815,10 +1817,10 @@ def test_device_update_boot_order(disk_type, disk_path, error_msg):
             vm.wait_for_login()
         # If we testing hotplug, detach the disk at last.
-        if device_at_dt_disk:
+        if device_at_dt_disk or device_cold_dt:
```
this is generic logic  for hotplug , suggest not mess up with device_cold_dt

dzhengfy replied:
Update

chunfuwen commented on libvirt/tests/src/virtual_disks/virtual_disks_multidisks.py:
```diff
@@ -1840,13 +1842,21 @@ def _check_disk_detach():
+
             # Check disks in VM after hotunplug.
-            if check_partitions_hotunplug:
+            if check_partitions_hotunplug and '--config' not in dt_options:
```
Be careful add "and"  here, that may impact other case to enter this conditional judgement

dzhengfy replied:
Update

chunfuwen commented on libvirt/tests/src/virtual_disks/virtual_disks_multidisks.py:
```diff
@@ -1840,13 +1842,21 @@ def _check_disk_detach():
                 if virtio_disk_hot_unplug_event_watch:
                     check_info_in_libvird_log_file('"event": "DEVICE_DELETED"')
+
+            if device_cold_dt:
```
I assume we can use currently check_vm_partitions to achieve current goal

chunfuwen commented on libvirt/tests/src/virtual_disks/virtual_disks_multidisks.py:
```diff
@@ -1840,13 +1842,21 @@ def _check_disk_detach():
                 if virtio_disk_hot_unplug_event_watch:
                     check_info_in_libvird_log_file('"event": "DEVICE_DELETED"')
+
+            if device_cold_dt:
```
If you want to do cold detach checking suggest move those checking to around line 1425  before vm.start

        vmxml = vm_xml.VMXML.new_from_dumpxml(vm_name)
        # Start the VM.
        logging.debug("Starting VM: %s", vmxml)
        vm.start()

dzhengfy replied:
Your suggestion does not satisfy the script requirement. I need to start vm, then cold unplug the disk, then check.

chunfuwen commented on libvirt/tests/src/virtual_disks/virtual_disks_multidisks.py:
```diff
@@ -1840,13 +1842,21 @@ def _check_disk_detach():
                         return False
                 # If disk_detach_error is False, then wait for seconds to let detach operation accomplish complete
-                if not disk_detach_error:
+                if not disk_detach_error and not device_cold_dt:
```
this additional condition is not necessary since cold plug/unplug will not enter this logic

dzhengfy replied:
Update

chunfuwen commented on libvirt/tests/src/virtual_disks/virtual_disks_multidisks.py:
```diff
@@ -1815,10 +1817,10 @@ def test_device_update_boot_order(disk_type, disk_path, error_msg):
             for i in list(range(len(disks))):
                 dt_options = ""
-                if devices[i] == "cdrom":
+                if devices[i] == "cdrom" or device_cold_dt:
```
this large block is for hotplug,  not mess up with device_cold_dt

dzhengfy replied:
Update

chunfuwen commented:
Can you upload test results after code updated due to addressed comments ?

dzhengfy commented:
> Can you upload test results after code updated due to addressed comments ?

@chunfuwen , sorry I did not make it clear. I already updated the test result in the 
https://github.com/autotest/tp-libvirt/pull/2654#issuecomment-610908487

This was edited after the codes are modified.

dzhengfy commented:
```
# date
Mon May 11 10:43:50 CST 2020
# avocado run --vt-type libvirt virtual_disks.multidisks..single_disk_test..disk_driver_option
JOB ID     : c54efd7faf5cacdf4abf1eb1429ce133fbe006a3
JOB LOG    : /root/avocado/job-results/job-2020-05-11T10.38-c54efd7/job.log
 (1/5) type_specific.io-github-autotest-libvirt.virtual_disks.multidisks.hotplug.single_disk_test.disk_driver_option.option_copy_on_read_on: PASS (59.78 s)
 (2/5) type_specific.io-github-autotest-libvirt.virtual_disks.multidisks.coldplug.single_disk_test.disk_driver_option.option_discard_unmap: PASS (51.76 s)
 (3/5) type_specific.io-github-autotest-libvirt.virtual_disks.multidisks.coldplug.single_disk_test.disk_driver_option.option_discard_ignore: PASS (52.04 s)
 (4/5) type_specific.io-github-autotest-libvirt.virtual_disks.multidisks.coldplug.single_disk_test.disk_driver_option.option_copy_on_read_off: PASS (53.08 s)
 (5/5) type_specific.io-github-autotest-libvirt.virtual_disks.multidisks.coldplug.single_disk_test.disk_driver_option.option_copy_on_read_on: PASS (50.30 s)
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 276.92 s

```
